### PR TITLE
EFF subdirectories

### PR DESCRIPTION
**Pull request by ln-zookeeper**
This allows animation frames of eff animations to be located in a subdirectory. This is useful for de-cluttering the effects directory in particular.

Usage and limitations as per http://www.hard-light.net/forums/index.php?topic=88139.0

Before merging, this needs testing in different mod environments: VP's, multiple mods, replacing anims from retail or other mods, and so on. I've pretty much only tested this in a development setting, as in a single mod as a bare directory, no VP's.

Notes:

I'm not sure about the original rationale behind the changes to the bm_unlock()/bm_unload() calls in bmpman.cpp; I'm sure they were intended as bugfixes, but I don't recall the specifics. I included them now because they've been like this in FotG for a long time without problems, but I'll have to split those into a separate commit or remove entirely before merging.

### CODE REVIEW

File: code/bmpman/bmpman.cpp [hunk 1]
```diff
@@ -45,7 +46,20 @@
 /**
  * @todo upgrade this to an inline funciton, taking bitmap_entry and const char* as arguments
  */
-#define EFF_FILENAME_CHECK { if ( be->type == BM_TYPE_EFF ) strcpy_s( filename, be->info.ani.eff.filename ); else strcpy_s( filename, be->filename ); }
+#define EFF_FILENAME_CHECK \
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: This macro is getting a bit too big, how about converting it to a static function?

File: code/bmpman/bmpman.cpp [hunk 2]
```diff
@@ -1506,8 +1539,8 @@ bitmap * bm_lock(int handle, ubyte bpp, ubyte flags, bool nodebug) {
 	// read the file data
 	if (gr_bm_lock(be->filename, handle, bitmapnum, bpp, flags, nodebug) == -1) {
 		// oops, this isn't good - reset and return NULL
-		bm_unlock(bitmapnum);
-		bm_unload(bitmapnum);
+		bm_unlock( handle );
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: This looks like the correct way of using the handle. `bitmapnum` should not be passed to the `bm_*` functions.
Could you submit this as a separate PR as it's not really part of the EFF subdirectories patch?

File: code/bmpman/bmpman.cpp (hunk 1 above)
Comment (ln-zookeeper) [inline code/bmpman/bmpman.cpp]: IIRC I didn't because it was the more complicated route and I didn't want to bundle potentially error-prone refactoring like that with the feature. It could certainly be done separately.

File: code/bmpman/bmpman.cpp [hunk 3]
```diff
@@ -2076,6 +2109,61 @@ void bm_page_in_start() {
...
+	mprintf(("BMPMAN: Setting temporary pathtype for %s... ", filename));
+
+	char *trimmed_path; // The differing portion of fullpath and fullrootpath
+	trimmed_path = &fullpath[strlen(fullrootpath)];
+
+	char *extpos = strstr(trimmed_path, "."); // Where the file extension begins
+	char basepath[MAX_PATH]; // This will hold the final, relative path
+
+	// Copy the relative path of the given file to basepath, excluding the extension
+	strncpy(basepath, trimmed_path, extpos-trimmed_path);
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: If `filename` does not have an extension then `extpos-trimmed_path` will be an invalid value. The function should either validate the return value of `strstr` or use an Assertion to make sure that no invalid values are passed to the function.

File: code/bmpman/bmpman.cpp [hunk 4]
```diff
@@ -2076,6 +2109,61 @@ void bm_page_in_start() {
...
+
+	// Make sure it's null-terminated right
+	basepath[(int)(extpos-trimmed_path)] = '\0';
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: There's no need to cast this to an int, a pointer difference is a valid array index (if the expression is actually valid, see above).

File: code/bmpman/bmpman.cpp [hunk 5]
```diff
@@ -2094,13 +2182,19 @@ void bm_page_in_stop() {
...
+					}
+
 					if (!gr_preload(bm_bitmaps[i].handle, (bm_bitmaps[i].preloaded == 2))) {
 						mprintf(("Out of VRAM.  Done preloading.\n"));
 						bm_preloading = 0;
 					}
 				} else {
 					bm_lock(bm_bitmaps[i].handle, (bm_bitmaps[i].used_flags == BMP_AABITMAP) ? 8 : 16, bm_bitmaps[i].used_flags);
-					bm_unlock(bm_bitmaps[i].handle);
+					if (bm_bitmaps[i].ref_count >= 1) {
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: If I understand the code of `bm_lock` correctly then this expression should always be true. `bm_lock` increments the ref_count so it is always at least 1 (if the function is successful).

File: code/bmpman/bmpman.cpp [hunk 6]
```diff
@@ -2076,6 +2109,61 @@ void bm_page_in_start() {
...
+
+	// Make sure it's null-terminated right
+	basepath[(int)(extpos-trimmed_path)] = '\0';
+
+	// Free the old string first, if it's not the const initial empty string
+	if (Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path[0] != '\0') {
+		vm_free(Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path);
+	}
+
+	Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path = vm_strdup(basepath);
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: The memory from this `vm_strdup` call is leaked at the end of the program. Previous values are freed but the last value that is allocated is not freed anywhere.
It's also a bit weird to mix string literals and heap-allocated string. At first `path` is assigned the empty string literal but here you assign a heap allocated string to `path`.

**Comment by ln-zookeeper** [comment]
The above feedback is now addressed as best I could. I ended up keeping the default value as an empty string literal instead of NULL because the latter would have then required adding a whole lot of NULL checks all over the place and that would have gotten messy and probably more bug-prone too.

In the past year the added APNG support may have rendered this maybe not quite as vital as before, but I think it's still a worthwhile feature since it can be used with any image format and allows existing animations to be compartmentalized more easily than converting them to APNG.

File: code/bmpman/bmpman.cpp [hunk 7]
```diff
@@ -1350,6 +1367,10 @@ bool bm_load_and_parse_eff(const char *filename, int dir_type, int *nframes, int
 		return false;
 	}
 
+	if (optional_string( "$Subdir:" )) {
+		stuff_boolean(&subdir);}
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: This brace is not aligned correctly.

File: code/bmpman/bmpman.cpp [hunk 8]
```diff
@@ -2576,6 +2609,61 @@ void bm_page_in_start() {
...
+
+	cf_create_default_path_string(fullrootpath, sizeof(fullrootpath) - 1, CF_TYPE_ROOT, NULL);
+	if (!cf_find_file_location(filename, CF_TYPE_ANY, sizeof(fullpath)-1, fullpath, &size, &offset)) {
+		mprintf(("BMPMAN: Failed to set temporary pathtype for %s: file not found!\n", filename));
+		return false;
+	}
+
+#ifndef NDEBUG
+	if (strlen(fullpath) < strlen(fullrootpath)) {
+		mprintf(("BMPMAN: Path to %s was shorter than the root path, programmer thought it could never happen!\n%s\n%s\n", filename, fullpath, fullrootpath));
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: If this should never happen then an `Assertion` might be the better choice here.

File: code/cfile/cfile.cpp [hunk 9]
```diff
@@ -235,6 +236,9 @@ int cfile_init(const char *exe_dir, const char *cdrom_dir)
 		Cfile_block_list[i].type = CFILE_BLOCK_UNUSED;
 	}
 
+	// Init to an empty string to avoid having to test for NULL everywhere
+	Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path = vm_strdup("");
```
Comment (asarium) [inline code/cfile/cfile.cpp]: This memory is leaked. This can be seen in the valgrind results of the unit tests.

File: code/bmpman/bmpman.cpp [hunk 10]
```diff
@@ -2576,6 +2609,61 @@ void bm_page_in_start() {
...
 }
 
+/**
+ * Stuffs Pathtypes with an additional path leading to a subdir of the same name as the given file (minus file extension).
+ * Used for allowing frames of .eff animations to be found in their own subdirectories.
+ *
+ * Example: if called with "debris.eff" and that file is in data/maps, then Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path
+ * will be stuffed with data/maps/debris, allowing data/maps/debris/debris_0000.dds (etc) to be found
+ */
+bool set_temp_subdir_pathtype(const char *filename) {
```
Comment (asarium) [inline code/bmpman/bmpman.cpp]: Could this function be expanded so that it would allow using a subdirectory in an arbitrary path type? This is currently tied to a specific file which means that it's very limited in what it can do.

**Comment by asarium** [comment]
Also, unit tests that check if this is actually working correctly would be nice.

File: code/bmpman/bmpman.cpp [hunk 11]
```diff
@@ -3151,6 +3243,8 @@ void bm_unload_all() {
 	}
 
 	Bm_ignore_load_count = 0;
+
+	vm_free(Pathtypes[CF_TYPE_TEMP_SUBDIR_LOOKUP].path);
```
Comment (MageKing17) [inline code/bmpman/bmpman.cpp]: Shouldn't this be in `cfile_close()` since the initial `vm_strdup()` call is in `cfile_init()`?

File: code/bmpman/bmpman.cpp (hunk 10 above)
Comment (ln-zookeeper) [inline code/bmpman/bmpman.cpp]: I'm not sure what exactly you're suggesting; as far as I can tell, there's not really any expanding to do and this should work as-is for any other files. Of course it will still be limited to anchoring the subdirectory to an existing filename (effects/foo.eff <-> effects/foo/) but that's not something I think I could get rid of.

If for some reason one needed to use it for non-bitmaps, it would be easy at that point to move the function to cfile and remove the references to BMPMAN from the output strings. However, to me that seems like a rather unlikely scenario.

Comment (asarium) [inline code/bmpman/bmpman.cpp]: This was more of a "nice-to-have" request so it's not very important. However, I just noticed that this function is defined in `bmpman.cpp` and not in the CFile system. The functionality of this function should be present in the CFile system and not in bmpman since it changes stuff inside the CFile system.

**Comment by ln-zookeeper** [comment]
Latest comments should again be addressed now. Except, of course, no unit tests since I have absolutely zero experience and knowledge about them.

**Comment by asarium** [comment]
I can write the unit tests, it should be relatively easy to do.

Have you tested this with VPs? I know that the cfile code does some indexing when loading a VP but I don't know if that includes all subdirectories or only the ones that are in the pathtypes.

**Comment by asarium** [comment]
I have added unit tests for testing the function: https://github.com/ln-zookeeper/fs2open.github.com/pull/3

However, I discovered that the function does not support VP files properly. The reason for this is that `cf_find_file_location` returns the path to the VP file if the requested file has been found in it which screws up where the subdirectory lookup searches for the other files.

**Comment by ln-zookeeper** [comment]
Files from VP's should work correctly now.

File: code/cfile/cfile.cpp [hunk 12]
```diff
@@ -1120,6 +1126,45 @@ int cf_get_dir_type(CFILE *cfile)
...
+	// (without extension) to it, giving us a relative path that can be temporarily
+	// stuffed into Pathtypes
+	for (size_t i=CF_TYPE_ROOT; i<CF_MAX_PATH_TYPES; i++) {
+		size_t size, offset; // Needed for cf_find_file_location, but not used
+
+		if (cf_find_file_location(filename, i, 0, NULL, &size, &offset)) {
+			SCP_string subdir, finalpath;
+
+			subdir = SCP_string(filename);
+			subdir = subdir.substr(0, subdir.find_last_of("."));
```
Comment (asarium) [inline code/cfile/cfile.cpp]: If filename doesn't contain an extension then will `find_last_of` will return a value that is not valid for `substr`.

Comment (ln-zookeeper) [inline code/cfile/cfile.cpp]: I didn't specifically test it, but according to documentation it'll return string::npos meaning that substr will just return the complete string.

Comment (asarium) [inline code/cfile/cfile.cpp]: Oh, well that's convenient.

File: code/cfile/cfile.cpp [hunk 13]
```diff
@@ -1120,6 +1126,45 @@ int cf_get_dir_type(CFILE *cfile)
...
+ * will be stuffed with data/maps/debris, allowing data/maps/debris/debris_0000.dds (etc) to be found
+ */
+bool cf_set_temp_subdir_pathtype(const char *filename) {
+	mprintf(("CFILE: Setting temporary pathtype for %s... ", filename));
+
+	// Finds the relative path of where the file exists and appends the filename
+	// (without extension) to it, giving us a relative path that can be temporarily
+	// stuffed into Pathtypes
+	for (size_t i=CF_TYPE_ROOT; i<CF_MAX_PATH_TYPES; i++) {
+		size_t size, offset; // Needed for cf_find_file_location, but not used
```
Comment (MageKing17) [inline code/cfile/cfile.cpp]: Er... if you don't need the size or the offset, you can just pass null pointers, like you already are for `pack_filename`.

Comment (ln-zookeeper) [inline code/cfile/cfile.cpp]: You're right. No idea why I had originally done it that way.

**Comment by MageKing17** [comment]
Given that we have APNGs, should this PR be closed?

**Comment by wookieejedi** [comment]
> Given that we have APNGs, should this PR be closed?

I would say yes?

**Comment by chief1983** [comment]
Are there not still performance advantages to using EFFs of DDS files vs APNGs?  As long as there are reasons to still use EFFs going forward, and therefore a use for this feature, it would still be nice to see merged at some point, I would think.

**Comment by wookieejedi** [comment]
> Are there not still performance advantages to using EFFs of DDS files vs APNGs? As long as there are reasons to still use EFFs going forward, and therefore a use for this feature, it would still be nice to see merged at some point, I would think.

Ah that's a good point, especially since DDS images can be mipmapped and loaded faster.

**Comment by wookieejedi** [comment]
Closing now that #4771 is merged.
